zebra-chain/src/sapling/tree.rs

```diff
@@ -72,37 +82,181 @@ impl fmt::Debug for Root {
     }
 }
 
-impl From<SaplingNoteCommitmentTree> for Root {
-    fn from(_tree: SaplingNoteCommitmentTree) -> Self {
-        // TODO: The Sapling note commitment tree requires a Pedersen
-        // hash function, not SHA256.
-
-        // let mut hash_writer = Sha256dWriter::default();
-        // sapling_note_commitment_tree
-        //     .zcash_serialize(&mut hash_writer)
-        //     .expect("A Sapling note commitment tree must serialize.");
-        // Self(hash_writer.finish())
+/// Sapling Note Commitment Tree
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+struct NoteCommitmentTree {
+    /// The root node of the tree (often used as an anchor).
+    root: Root,
+    /// The height of the tree (maximum height for Sapling is 32).
+    height: u8,
+    /// The number of leaves (note commitments) in this tree.
+    count: u32,
+}
 
+impl From<Vec<NoteCommitment>> for NoteCommitmentTree {
+    fn from(_values: Vec<NoteCommitment>) -> Self {
         unimplemented!();
     }
 }
 
-impl SaplingNoteCommitmentTree {
-    /// Get the Jubjub-based Pedersen hash of root node of this merkle
-    /// tree of commitment notes.
-    pub fn hash(&self) -> [u8; 32] {
-        unimplemented!();
+impl From<Vec<jubjub::Fq>> for NoteCommitmentTree {
+    fn from(values: Vec<jubjub::Fq>) -> Self {
+        if values.is_empty() {
+            return NoteCommitmentTree {
+                root: Root::default(),
+                height: 0,
+                count: 0,
+            };
+        }
+
+        let count = values.len() as u32;
+        let mut height = 0u8;
+        let mut current_layer: Vec<[u8; 32]> =
+            values.into_iter().map(|cm_u| cm_u.to_bytes()).collect();
+
+        while usize::from(height) < MERKLE_DEPTH {
+            let mut next_layer_up = vec![];
+
+            while !current_layer.is_empty() {
+                let left = current_layer.remove(0);
+                let right;
+                if current_layer.is_empty() {
+                    right = EMPTY_ROOTS[height as usize];
+                } else {
+                    right = current_layer.remove(0);
+                }
```

Comment:
TIOLI: since arrays are copy if their elements are you should be able to write it like this without any performance costs:
```suggestion
                let right = current_layer
                    .get(0)
                    .cloned()
                    .unwrap_or_else(|| EMPTY_ROOTS[height as usize]);
```

Reply:
I need to keep modifying `current_layer` not just copying/cloning elements from it, else it will loop forever

Reply:
If `Vec::remove` returned an `Option` rather than panicing, this could be written as `let right = current_layer.remove(0).unwrap_or_else(|| ...)`.

But `Vec::remove` isn't an ideal method to use if it's called as `Vec::remove(0)`, because it shifts all subsequent elements forward in memory.  Instead, [`VecDeque`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html) provides a ring buffer that can efficiently insert/remove from either end.  [`VecDeque::pop_front`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html#method.pop_front) looks like it might be a good fit here.

Reply:
Aha nice, I'll try that

Reply:
https://github.com/ZcashFoundation/zebra/pull/1045/commits/2d41993b7371c66c336003f0a1a6f0162e9f3599